comparison/comparison.go

```diff
@@ -0,0 +1,509 @@
+	}
+
+	switch lt.Type {
+	case LoadTestTypeBounded:
+		coordConfig.ClusterConfig.MaxActiveUsers = lt.NumUsers
+		// TODO: uncomment this line and remove the loop after a new release is
+		// published.
+		// coordConfig.MonitorConfig.Queries = nil
+		for i := 0; i < len(coordConfig.MonitorConfig.Queries); i++ {
+			coordConfig.MonitorConfig.Queries[i].Alert = false
+		}
+		duration, parseErr := time.ParseDuration(lt.Duration)
+		if parseErr != nil {
+			return status, parseErr
+		}
+		return runBoundedLoadTest(t, coordConfig, duration, cancelCh)
+	case LoadTestTypeUnbounded:
+		// TODO: cleverly set MaxActiveUsers to (numAgents * UsersConfiguration.MaxActiveUsers)
+		return runUnboundedLoadTest(t, coordConfig, cancelCh)
+	}
+
+	return status, fmt.Errorf("unimplemented LoadTestType %s", lt.Type)
+}
+
+func (c *Comparison) getResults(resultsCh <-chan Result) ([]Result, error) {
+	var results []Result
+	for res := range resultsCh {
+		dp := c.deployments[res.deploymentID]
+		t := terraform.New(res.deploymentID, &dp.config)
+		defer t.Cleanup()
+		output, err := t.Output()
+		if err != nil {
+			return results, err
+		}
+
+		promURL := "http://" + output.MetricsServer.PublicIP + ":9090"
+		helper, err := prometheus.NewHelper(promURL)
+		if err != nil {
+			return results, fmt.Errorf("failed to create prometheus.Helper: %w", err)
+		}
+		g := report.New(res.LoadTests[0].Label, helper, dp.config.Report)
+		baseReport, err := g.Generate(res.LoadTests[0].Status.StartTime, res.LoadTests[0].Status.StopTime)
+		if err != nil {
+			return results, fmt.Errorf("error while generating report: %w", err)
+		}
+		g = report.New(res.LoadTests[1].Label, helper, dp.config.Report)
+		newReport, err := g.Generate(res.LoadTests[1].Status.StartTime, res.LoadTests[1].Status.StopTime)
+		if err != nil {
+			return results, fmt.Errorf("error while generating report: %w", err)
+		}
+
+		if c.config.Output.GenerateReport {
+			var buf bytes.Buffer
+			opts := report.CompareOpts{
+				GenGraph:     c.config.Output.GenerateGraphs,
+				GraphsPrefix: fmt.Sprintf("%s%d_", res.deploymentID, res.LoadTests[0].loadTestID),
+			}
+			err := report.Compare(&buf, opts, baseReport, newReport)
+			if err != nil {
+				return results, err
+			}
+			res.Report = buf.String()
+		}
+
+		if c.config.Output.UploadDashboard {
+			var dashboardData bytes.Buffer
+			title := fmt.Sprintf("Comparison - %d - %s - %s",
+				res.LoadTests[0].loadTestID, res.LoadTests[0].Config.DBEngine, res.LoadTests[0].Config.Type)
+			if err := report.GenerateDashboard(title, baseReport, newReport, &dashboardData); err != nil {
+				return results, err
+			}
+
+			url, err := t.UploadDashboard(dashboardData.String())
+			if err != nil {
+				return results, err
+			}
+			res.DashboardURL = fmt.Sprintf("http://%s:3000%s", output.MetricsServer.PublicIP, url)
+		}
+
+		results = append(results, res)
+	}
+
+	return results, nil
+}
+
+func provisionBuilds(t *terraform.Terraform, baseBuildURL, newBuildURL string) error {
```

Comment:
Would be good to have a godoc on what this function does. IIUC, this uploads the base build and new build to all the servers? So the default terraform setup just downloads latest.mattermost.com?

Reply:
Sure, I'll add a few comments. This is either downloading (if it's a URL) or uploading (if prefixed by `file://`) pure builds (no binaries) as provided in the comparison configuration. I am doing this only once after everything has deployed use the files later on in `initLoadTest` where we do the actual installation.